crates/task-impls/src/helpers.rs

```diff
@@ -42,15 +42,38 @@ pub async fn broadcast_event<E: Clone + std::fmt::Debug>(event: E, sender: &Send
 ///
 /// # Panics
 /// Panics if the VID calculation fails, this should not happen.
+#[allow(clippy::panic)]
 pub async fn calculate_vid_disperse<TYPES: NodeType>(
     txns: Vec<u8>,
     membership: Arc<TYPES::Membership>,
     view: TYPES::Time,
 ) -> VidDisperse<TYPES> {
     let num_nodes = membership.total_nodes();
     let vid_disperse = spawn_blocking(move || {
-        #[allow(clippy::panic)]
-        vid_scheme(num_nodes).disperse(&txns).unwrap_or_else(|err|panic!("VID disperse failure:\n\t(num_storage nodes,payload_byte_len)=({num_nodes},{})\n\terror: : {err}", txns.len()))
+        vid_scheme(num_nodes).disperse(&txns).unwrap_or_else(|err| panic!("VID disperse failure:(num_storage nodes,payload_byte_len)=({num_nodes},{}) error: {err}", txns.len()))
+    })
+    .await;
+    #[cfg(async_executor_impl = "tokio")]
+    // Unwrap here will just propagate any panic from the spawned task, it's not a new place we can panic.
+    let vid_disperse = vid_disperse.unwrap();
+
+    VidDisperse::from_membership(view, vid_disperse, membership.as_ref())
+}
+
+/// Calculate the vid disperse information from the payload given a view and membership, and precompute data from builder
+///
+/// # Panics
+/// Panics if the VID calculation fails, this should not happen.
+#[allow(clippy::panic)]
+pub async fn calculate_vid_disperse_using_precompute_data<TYPES: NodeType>(
+    txns: Vec<u8>,
+    membership: Arc<TYPES::Membership>,
+    view: TYPES::Time,
+    pre_compute_data: VidPrecomputeData,
+) -> VidDisperse<TYPES> {
+    let num_nodes = membership.total_nodes();
+    let vid_disperse = spawn_blocking(move || {
```

Comment:
This one too...

Reply:
https://github.com/EspressoSystems/HotShot/pull/2956/commits/5408f863bc99171c7e299ac0460875145d98252d